lean/monad-stacks.lean

```diff
@@ -1,33 +1,250 @@
+Before you read this: This chapter does not attempt to introduce the concept
+of a monad in general, it assume this as a given and explains the monads
+and constructs on top of them that are essential to Lean meta programming.
+If you don't know what a monad is already you can read (TODO: Link to monad tutorial).
 
-# Monad transformers and Mondas relevant for meta
-* ReaderT
-* StateRefT
-
-# Monad extensions
-* MonadLift
-* MonadState
-* MonadExcept
-* MonadReader
-* MonadControl
-* MonadFinally
-* MonadFunctor
-* MonadStateOf
-* MonadControlT
-* MonadEnv
-* MonadRef
-* MonadMCtx
-* MonadLCtx
-* MonadTrace
-* MonadCache
-* MonadError
-* MonadWithReader
-* MonadOptions
-* MonadRecDepth
-* MonadBacktrack
-
-* MonadError
-* ...
-* `liftM`
+## Monad Transformers
+Quite often in functional programming with monads one wants to have the power
+of more than one monad available. Specifically in Lean meta programming it
+is a very common pattern that we have some sort of read only input, an
+execution context, available to us, as well as the ability to operate on
+a mutatable state. Both of these issues have monads that can solve them
+seperately already, namely the `Reader` monad for read only input and the
+`StateM` monad for the mutable state. Now we can try to just "stack" them
+ontop like this:`
+-/
+namespace Playground1
+abbrev Context := Nat
+abbrev State := String
+
+abbrev MyM (α : Type) := Reader Context (StateM State α)
+
+end Playground1
+/-!
+And this will indeed type check, however we have a very annoying issue:
+-/
+namespace Playground1
+
+#synth Monad MyM -- no monad instance :(
+
+end Playground1
+/-!
+We can fix this by manually declaring the instance and maybe even write functions
+that make `Reader` and `StateM` play nicely together but it would of course
+be much nicer if we could just have all of this done for us automatically,
+this is where monad transformers come in. A monad transformer gives us a way
+to combine the effects that certain monads give us (not all monad can be turned into
+a transformer) in a nice way. For these monads specifically there is the `Reader`
+transformer `ReaderT` and the `StateM` transformer `StateT`, let's see how they
+perform:
+-/
+namespace Playground2
+abbrev Context := Nat
+abbrev State := String
+
+-- `StateM` is defined as `StateM σ α = StateT σ Id α` so the following two are equivalent.
+abbrev MyM  := ReaderT Context (StateT State Id)
+abbrev MyM2  := ReaderT Context (StateM State)
+-- This definition makes perfect sense since the `Id` monad has no effect,
+-- can you guess the definition of `Reader`?
+
+#synth Monad MyM -- monad instance!
+#synth Monad MyM2 -- monad instance!
+
+end Playground2
+/-!
+So that's the monad instance for free. An other thing we would want is the
+ability to still easily `read`, `get` and `set`, after all this is the effects
+we just combined, let's see whether our monad transformers allow this:
+-/
+namespace Playground2
+
+def test : MyM String := do
```

Comment:
I feel like this test is doing two things at once. It's a bit hard to follow the idea at this point